### ISSUE TITLE
Let’s talk about change tracking and the specification tests

### ISSUE THREAD

**Issue by tuespetre** [dotnet/efcore#11513]
I’m working on integrating a custom `IQueryCompiler` with EF Core and it’s going pretty well. I’m finding some ‘internal’ APIs useful so I plan to create an issue for that later to discuss how I am using them and what should be formalized as ‘public’ or what should be used instead. In this issue, I would just like to discuss the specification tests and change tracking and how they come together.

A huge portion of these specification tests relies on assertions about what does **not** make it into the change tracker, and I feel that is problematic. I am all for performance optimizations, but searching through the issues in this repository I see some issues related to entities not being put into the change tracker when they should. For this exact reason I would rather my query compiler be extremely optimistic about what goes into the change tracker during a query’s execution and only apply optimizations to that with surgical precision. However, that huge portion of tests I mentioned all fail under this philosophy: they assert that **only** a select few entries should be present in the change tracker, rather than just asserting that some specific entities are present in the change tracker regardless of what else is present. Because of this I cannot reliably use the tests to ensure that my query compiler is accurately translating queries in accordance to EF Core specifications.

TL;DR: Can we have the ‘query’ specification tests verify that the resulting entities are present in the change tracker rather than verifying that only an exact number of entries is present in the change tracker? Perhaps it is the duty of a specific test class, and not all of the query test classes, to assert exact numbers related to optimizations thereof.

**Comment by tuespetre** [comment]
Addendum: I’ve been operating from `release/2.0` but I see that in 2.1 the query tests have been refactored in some pretty nice ways, where I could supply a custom QueryAsserter that handles the entry count differently. I think if I work from that angle instead I might be in the clear.

**Comment by ajcvickers** [comment]
@tuespetre Can you explain a bit about why it would be okay from a functional/correctness perspective for entities to be tracked when they are not returned from the query?

**Comment by tuespetre** [comment]
I don’t see an issue with functionality/correctness if some extra entities are added to the tracker, because I think that unless you are configuring the provider to throw on client evaluation, you are in a sense agreeing to the possibility of having extra entities tracked (any ‘selector’ that runs at the client may result in extra tracking entries.) 

On the other side of the coin, it is absolutely not okay to *not* track an entity when it is expected to be tracked.

**Comment by ajcvickers** [comment]
@tuespetre Thanks for the info. We will discuss as a team, but given that one reason for running a query is to track specifically the results of the query in, for example, data binding scenarios, I don't think we want to relax the contract and say that it's okay for additional entities to be tracked. 

Also note that selectors running in client eval may result in additional entities being materialized, but if these end up being tracked even though not actually returned from the query, then that would be a bug, and we should investigate.

**Comment by tuespetre** [comment]
The problem that arises from being too pessimistic about what to track can be seen here:

```
from c in customers 
from o1 in c.Orders
from o2 in c.Orders
where ClientPredicate(o1, o2)
select ClientSelector(o1, o2)
```

What does ClientSelector return? Maybe it’s choosing between one order or the other somehow, or maybe it’s returning some other object that includes them both, like an array or a DTO. No query provider can really be certain about whether it is okay to track one order or the other or both or neither, and while it should be no sin to optimistically track both, it would be undesired to not track either of them when either of them is needing to be tracked.

There are of course ‘workarounds’ in such scenarios, but they take the matter of the leaky abstraction from being a possible performance concern (that could be fine tuned later) to being a logical or behavioral problem where the authored query doesn’t ‘just work’ and produce the expected behavior.

**Comment by Wain123** [comment]
>it would be undesired to not track either of them when either of them is needing to be tracked

This was (unfortunately) marked as by-design in #10471

**Comment by ajcvickers** [comment]
@tuespetre In that case it would depend on the shape returned by ClientSelector--if it returns orders, then those would be tracked. If instead you want orders to be tracked that are not returned, then explicitly using the ClientSelector would be the way to get more tracked.

We discussed this in triage and we don't think that it makes sense to change the semantics of what gets tracked and what doesn't--it should remain that only entities returned by the query are tracked. That being said, if there is a small PR that allows you to reuse the tests with different semantics, then we would consider taking it.

@Wain123 This is not the same issue as discussed in #10471. That was about when entities are tracked, not whether or not they are tracked.

**Comment by tuespetre** [comment]
@ajcvickers that is fair enough. I agree that ultimately only returned entities should be tracked as well. With the way the tests have been refactored from 2.0 to 2.1, it could be as simple as introducing a virtual flag property on `QueryTestBase` that controls whether or not to verify the entry count. 

I also found a couple of relevant scenarios where we differ:

- `GroupBy_with_element_selector2`
- `GroupBy_with_element_selector3`

In both of these tests, I am pulling the entire grouping out from the query (making use of `FOR JSON` in SQL Server) and tracking the entities within the grouping immediately, whereas EF Core's default query compiler uses a custom grouping class that causes tracking only upon iteration.

**Comment by tuespetre** [comment]
@ajcvickers alright, so I have implemented my own 'tracking pruning' and it is for the most part true to what EF Core already does. I did uncover some main groups of tests that I had to skip:

- Tests where my compiler is tracking more entities than the test expects, and looking at the queries defined by the EF Core tests, it seems like a bug that EF Core is not tracking those entities. **There are 56 of these tests.**

    Example:
    https://github.com/aspnet/EntityFrameworkCore/blob/28479da1f8986a88fc8a067ceab1b9ea45afb637/src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.JoinGroupJoin.cs#L187-L196

- Tests where the query makes use of an intermediate navigation in the projection and it causes EF Core to track that intermediate navigation even though it is ultimately not a part of the projection. My compiler is not tracking them. I am not sure if this is a bug with EF Core or not. **There are 2 of these tests.**

    Example:

    https://github.com/aspnet/EntityFrameworkCore/blob/28479da1f8986a88fc8a067ceab1b9ea45afb637/src/EFCore.Specification.Tests/Query/QueryNavigationsTestBase.cs#L422-L436


- Tests where the results are assumed to be in a certain order even though there is no semantic reason for the query to be ordered that way. From what I recall these are all related to how EF Core optimizes its client `GroupJoin` handling. These tests could be tweaked to supply an `elementSorter` so that becomes invariant. **There are 11 of these tests.** 
- Tests where we disagree about whether or not the entities should be tracked because they pass through a client selector. **There are 4 of these tests.**
- Tests where my compiler is using `FOR JSON` to serialize and pull out `IGrouping<,>` and `IEnumerable<>` nested collections and track them right away. I have no complaints about EF Core's approach to tracking those, this is just a feature difference. For completeness: **There are 8 of these tests.**

**Comment by smitpatel** [comment]
For pt. 2 above, the test posted is tracking 30 entities which seems correct no. given there are 30 orders on the server which have CustomerId starting with 'A'. Since Customer is not being projected out, it should not get tracked.

**Comment by tuespetre** [comment]
You are right. I am on the 2.0 release branch and it has 34 so that must be corrected for 2.1. 🎉

**Comment by tuespetre** [comment]
I’m going to go ahead and close this. Turns out it’s just easier to copy the original test’s source in an override and tweak the entry count or ordering that way. A couple of tests in this repository even do that: something with string.Contains if I remember correctly.

Also, since working off of the 2.1 branch, that first class of tests I described above is down from 56 to 7 or 8. Good to see that!
